### PR TITLE
DOC: Rewritten style.py docstrings: #12148

### LINKED ISSUE #12148: Docstring on Styler.apply{,map} is confusing
[Right now it says](https://github.com/pydata/pandas/blob/564166529071d33844712ab6038f6a237ddcb8f1/pandas/core/style.py#L357):

``` python
    def apply(self, func, axis=0, subset=None, **kwargs):
        """
        Apply a function column-wise, row-wise, or table-wase,
        updating the HTML representation with the result.
        .. versionadded:: 0.17.1
        Parameters
        ----------
        func: function
        axis: int, str or None
            apply to each column (``axis=0`` or ``'index'``)
            or to each row (``axis=1`` or ``'columns'``) or
            to the entire DataFrame at once with ``axis=None``.
        subset: IndexSlice
            a valid indexer to limit ``data`` to *before* applying the
            function. Consider using a pandas.IndexSlice
        kwargs: dict
            pass along to ``func``
        Returns
        -------
        self
        Notes
        -----
        This is similar to ``DataFrame.apply``, except that ``axis=None``
        applies the function to the entire DataFrame at once,
        rather than column-wise or row-wise.
        """
```

I spent ages trying to figure out how to feed any kind of function into this, only to eventually find the [Styler's docstring](https://github.com/pydata/pandas/blob/564166529071d33844712ab6038f6a237ddcb8f1/pandas/core/style.py#L68-L77):

``` python
    """
    Notes
    -----
    Most styling will be done by passing style functions into
    ``Styler.apply`` or ``Styler.applymap``. Style functions should
    return values with strings containing CSS ``'attr: value'`` that will
    be applied to the indicated cells.
    If using in the Jupyter notebook, Styler has defined a ``_repr_html_``
    to automatically render itself. Otherwise call Styler.render to get
    the genterated HTML.
    """
```

Really, the docstring of apply should state that the passed function should return CSS to be applied to the element. Otherwise, it just sounds like something which modifies the data before it is presented.

My use case is that I have a column full of [`uncertainties.ufloat`](https://pythonhosted.org/uncertainties/user_guide.html), and I want to convert them to a particular convenient string representation before they are formatted in a HTML table. I still can't figure out how to cleanly do this...

### PR DESCRIPTION
**Pull request by aaronmichaelsilberstein**
Docstrings for pandas/io/formats/style.py now accurately reflects the behavior of the ``Styler`` class. Fixed a typo in the docstring of the `style` property of ``DataFrame``.

- [x] closes #12148 
- [ ] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/io/formats/style.py [hunk 1]
```diff
@@ -49,19 +48,21 @@ def _mpl(func):
 
 class Styler(object):
     """
-    Helps style a DataFrame or Series according to the
-    data with HTML and CSS.
+    A decorator class for a `DataFrame` object to provide an
```
Comment (TomAugspurger) [inline pandas/io/formats/style.py]: What's a "decorator class"? This like it'd be easily confused with python decorators, which I don't think you mean.

File: pandas/io/formats/style.py [hunk 2]
```diff
@@ -72,32 +73,46 @@ class Styler(object):
...
-      * ``level<k>`` where `k` is the level in a MultiIndex
-
-    * Column label cells include
-      * ``col_heading``
-      * ``col<n>`` where `n` is the numeric position of the column
-      * ``evel<k>`` where `k` is the level in a MultiIndex
-
-    * Blank cells include ``blank``
-    * Data cells include ``data``
+    `Styler` renders HTML on a `DataFrame` using its own `Styler.render`
```
Comment (TomAugspurger) [inline pandas/io/formats/style.py]: I don't follow "renders HTML on a DataFrame". Perhaps "renders an HTML representation of a DataFrame"?

File: pandas/io/formats/style.py [hunk 3]
```diff
@@ -72,32 +73,46 @@ class Styler(object):
...
+    method; in particular, the Styler class will not affect, and is
+    independent from, its `data` object's `DataFrame.to_html` rendering
+    method.
+
+    `Styler.render` returns an HTML string, based on attributes which are set
```
Comment (TomAugspurger) [inline pandas/io/formats/style.py]: Don't need this comma.

File: pandas/io/formats/style.py [hunk 4]
```diff
@@ -72,32 +73,46 @@ class Styler(object):
 
     Notes
     -----
-    Most styling will be done by passing style functions into
-    ``Styler.apply`` or ``Styler.applymap``. Style functions should
-    return values with strings containing CSS ``'attr: value'`` that will
-    be applied to the indicated cells.
 
-    If using in the Jupyter notebook, Styler has defined a ``_repr_html_``
```
Comment (TomAugspurger) [inline pandas/io/formats/style.py]: Why did you remove this? I think it's helpful to note that the class defines `_repr_html_`.

File: pandas/io/formats/style.py [hunk 5]
```diff
@@ -764,22 +833,40 @@ def set_caption(self, caption):
...
 
         Returns
         -------
         self : Styler
 
+        Notes
+        -----
+        Entries are rendered with CSS selectors according to gross properties
+        of their position and content:
+        * Index and Column names include ``index_name`` and ``level<k>``
```
Comment (TomAugspurger) [inline pandas/io/formats/style.py]: `set_table_styles` doesn't seem like the appropriate place for this section. Perhaps `.render`?

File: pandas/io/formats/style.py [hunk 6]
```diff
@@ -813,7 +900,7 @@ def hide_columns(self, subset):
         Parameters
         ----------
         subset: IndexSlice
-            An argument to ``DataFrame.loc`` that identifies which columns
```
Comment (TomAugspurger) [inline pandas/io/formats/style.py]: @jorisvandenbossche Do we have an official policy on single vs. double backticks? I know numpydoc recommends single, but I thought we used double so that it looked nicer in the HTML docs.

File: doc/source/whatsnew/v0.22.0.txt [hunk 7]
```diff
@@ -250,8 +250,7 @@ Performance Improvements
 
 Documentation Changes
 ~~~~~~~~~~~~~~~~~~~~~
-
--
+- Heredoc documentation for io/formats/style.py has been overhauled to more accurately reflect the functioning of the `Styler` class. (:issue:`12148`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.22.0.txt]: What's a "Heredoc"?

**Comment by TomAugspurger** [comment]
Thanks, looks like a nice improvement. Just a few questions.

File: pandas/io/formats/style.py (hunk 1 above)
Comment (aaronmichaelsilberstein) [reply pandas/io/formats/style.py]: By "decorator class" I mean the concept precisely defined on p. 175 of "Design Patterns" by Gamma et al (commonly referred to as Gang of Four, or GOF), which seems to precisely describe what Styler is: a class which contains additional state information to perform an auxiliary function on another class.  Is there a name we consistently use in our documentation to refer to such a concept?

File: pandas/io/formats/style.py (hunk 4 above)
Comment (aaronmichaelsilberstein) [reply pandas/io/formats/style.py]: I didn't remove it - it's now in line 109.

**Review by aaronmichaelsilberstein — COMMENTED** [review]
Thanks for your careful review!

**Comment by jreback** [comment]
closing as stale. ping if you want to update.
